### PR TITLE
Add paint by example

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds https://github.com/Fantasy-Studio/Paint-by-Example 

- [x] Weight Conversion 

**Conversion script works**:
```bash
python scripts/convert_original_stable_diffusion_to_diffusers.py --checkpoint_path ../new_inpaint_test/model.ckpt --dump_path ../new_inpaint_test --num_in_channels 9 --pipeline_type PaintByExample
```
**Test weights are uploaded**
```python
from diffusers import DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained("Fantasy-Studio/Paint-by-Example")
```

- [x] Check forward pass - integration test works as expected
- [x] Demo
- [ ] Publish

Check with authors: @Fantasy-Studio

### CODE REVIEW

File: src/diffusers/models/attention.py [hunk 1]
```diff
@@ -310,6 +310,7 @@ def set_use_memory_efficient_attention_xformers(self, use_memory_efficient_atten
                 )
             except Exception as e:
                 raise e
+
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention.py:313]:
```suggestion
```

File: tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py [hunk 2]
```diff
@@ -167,8 +167,8 @@ def test_stable_diffusion_inpaint(self):
         tokenizer = CLIPTokenizer.from_pretrained("hf-internal-testing/tiny-random-clip")
 
         image = self.dummy_image.cpu().permute(0, 2, 3, 1)[0]
-        init_image = Image.fromarray(np.uint8(image)).convert("RGB").resize((64, 64))
         mask_image = Image.fromarray(np.uint8(image + 4)).convert("RGB").resize((64, 64))
+        init_image = Image.fromarray(np.uint8(image)).convert("RGB").resize((64, 64))
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py:171]:
```suggestion
```

File: docs/source/api/pipelines/overview.mdx [hunk 3]
```diff
@@ -53,6 +53,7 @@ available a colab notebook to directly try them out.
 | [latent_diffusion](./api/pipelines/latent_diffusion) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752)| Text-to-Image Generation | 
 | [latent_diffusion](./api/pipelines/latent_diffusion) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752)| Super Resolution Image-to-Image | 
 | [latent_diffusion_uncond](./api/pipelines/latent_diffusion_uncond) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752) | Unconditional Image Generation | 
+| [paint_by_example](./api/pipelines/paint_by_example) | [**Paint by Example: Exemplar-based Image Editing with Diffusion Models**](https://arxiv.org/abs/2211.13227) | Image Guided Image Inpainting |
```
Comment (pcuenca) [inline docs/source/api/pipelines/overview.mdx:56]:
```suggestion
| [paint_by_example](./api/pipelines/paint_by_example) | [**Paint by Example: Exemplar-based Image Editing with Diffusion Models**](https://arxiv.org/abs/2211.13227) | Image-Guided Image Inpainting | 
```

File: docs/source/index.mdx [hunk 4]
```diff
@@ -43,6 +43,7 @@ available a colab notebook to directly try them out.
 | [latent_diffusion](./api/pipelines/latent_diffusion) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752)| Text-to-Image Generation | 
 | [latent_diffusion](./api/pipelines/latent_diffusion) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752)| Super Resolution Image-to-Image | 
 | [latent_diffusion_uncond](./api/pipelines/latent_diffusion_uncond) | [**High-Resolution Image Synthesis with Latent Diffusion Models**](https://arxiv.org/abs/2112.10752) | Unconditional Image Generation | 
+| [paint_by_example](./api/pipelines/paint_by_example) | [**Paint by Example: Exemplar-based Image Editing with Diffusion Models**](https://arxiv.org/abs/2211.13227) | Image Guided Image Inpainting |
```
Comment (pcuenca) [inline docs/source/index.mdx:46]:
```suggestion
| [paint_by_example](./api/pipelines/paint_by_example) | [**Paint by Example: Exemplar-based Image Editing with Diffusion Models**](https://arxiv.org/abs/2211.13227) | Image-Guided Image Inpainting | 
```

File: src/diffusers/models/attention.py [hunk 5]
```diff
@@ -415,23 +418,28 @@ def __init__(
...
-            self.norm1 = nn.LayerNorm(dim)
-            self.norm2 = nn.LayerNorm(dim)
+            self.attn2 = None
+
+        self.norm1 = AdaLayerNorm(dim, num_embeds_ada_norm) if self.use_ada_layer_norm else nn.LayerNorm(dim)
+
+        if cross_attention_dim is not None:
+            self.norm2 = AdaLayerNorm(dim, num_embeds_ada_norm) if self.use_ada_layer_norm else nn.LayerNorm(dim)
+        else:
+            self.attn2 = None
```
Comment (pcuenca) [inline src/diffusers/models/attention.py:440]:
```suggestion
            self.norm2 = None
```

(I think)

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+from ..stable_diffusion.safety_checker import StableDiffusionSafetyChecker
+from .image_encoder import PaintByExampleImageEncoder
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def prepare_mask_and_masked_image(image, mask):
+    """
+    Prepares a pair (image, mask) to be consumed by the Stable Diffusion pipeline. This means that those inputs will be
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:39]:
```suggestion
    Prepares a pair (image, mask) to be consumed by the Paint by Example pipeline. This means that those inputs will be
```

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+            if mask.shape[0] == 1:
+                mask = mask.unsqueeze(0)
+
+            # Batched masks no channel dim
+            else:
+                mask = mask.unsqueeze(1)
+
+        assert image.ndim == 4 and mask.ndim == 4, "Image and Mask must have 4 dimensions"
+        assert image.shape[-2:] == mask.shape[-2:], "Image and Mask must have the same spatial dimensions"
+        assert image.shape[0] == mask.shape[0], "Image and Mask must have the same batch size"
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:90]:
```suggestion
        assert image.shape[0] == mask.shape[0], "Image and Mask must have the same batch size"
        assert image.mask[1] == 1, "Mask image must have a single channel"
```

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+
+        # Batch single mask or add channel dim
+        if mask.ndim == 3:
+            # Single batched mask, no channel dim or single mask not batched but channel dim
+            if mask.shape[0] == 1:
+                mask = mask.unsqueeze(0)
+
+            # Batched masks no channel dim
+            else:
+                mask = mask.unsqueeze(1)
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:86]: This is a bit confusing to visualize. Would something like this be easier?

```suggestion
            # Batched mask
            if mask.shape[0] == image.shape[0]:
                 mask = mask.unsqueeze(1)
             else:
                 mask = mask.unsqueeze(0)
```

In any case we should check that the channel dim is `1` (assert added later).

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        mask[mask < 0.5] = 0
+        mask[mask >= 0.5] = 1
+
+        # Image as float32
+        image = image.to(dtype=torch.float32)
+    elif isinstance(mask, torch.Tensor):
+        raise TypeError(f"`mask` is a torch.Tensor but `image` (type: {type(image)} is not")
+    else:
+        if isinstance(image, PIL.Image.Image):
+            image = np.array(image.convert("RGB"))
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:113]: Assert image and mask are np arrays?

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 10]
```diff
@@ -0,0 +1,559 @@
...
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, torch.FloatTensor], None]] = None,
+        callback_steps: Optional[int] = 1,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
+
+        Args:
+            prompt (`str` or `List[str]`):
+                The prompt or prompts to guide the image generation.
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:388]:
```suggestion
            example_image (`torch.FloatTensor` or `PIL.Image.Image` or `List[PIL.Image.Image]`):
                The exemplar image to guide the image generation.
```

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+            image (`PIL.Image.Image`):
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:389]: Same types as `example_image`, I think: a single PIL image or a list, o tensor or a numpy array.

File: src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+                `Image`, or tensor representing an image batch which will be inpainted, *i.e.* parts of the image will
+                be masked out with `mask_image` and repainted according to `prompt`.
+            mask_image (`PIL.Image.Image`):
```
Comment (pcuenca) [inline src/diffusers/pipelines/paint_by_example/pipeline_paint_by_example.py:392]: Same

**Review by pcuenca — APPROVED** [review]
Very very cool!

Question: should we place the folder inside `pipelines/stable_diffusion` so it's a bit more visible?

File: src/diffusers/pipelines/paint_by_example/__init__.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+from dataclasses import dataclass
+from enum import Enum
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/paint_by_example/__init__.py:2]:
```suggestion
```

File: tests/test_pipelines_common.py [hunk 14]
```diff
@@ -302,7 +302,7 @@ def test_attention_slicing_forward_pass(self):
         output_with_slicing = pipe(**inputs)[0]
 
         max_diff = np.abs(output_with_slicing - output_without_slicing).max()
-        self.assertLess(max_diff, 1e-5, "Attention slicing should not affect the inference results")
+        self.assertLess(max_diff, 1e-3, "Attention slicing should not affect the inference results")
```
Comment (patrickvonplaten) [inline tests/test_pipelines_common.py:305]: attention slicing forces different batch size matrix multiplication - there is no way we can guarantee a 1e-5 diff only IMO

Comment (patrickvonplaten) [reply tests/test_pipelines_common.py:305]: cc @anton-l

**Comment by patrickvonplaten** [comment]
> Very very cool!
> 
> Question: should we place the folder inside `pipelines/stable_diffusion` so it's a bit more visible?

New paper -> new folder IMO

**Comment by chenbinghui1** [comment]
@patrickvonplaten hello, how to convert diffuser's model to original paintbyexample model type. I find there is no scripts for doing this

**Pull request merged**